src/sqswatcher/plugins/slurm.py

```diff
@@ -61,55 +64,120 @@ def _reconfigure_nodes():
     with tempfile.NamedTemporaryFile(mode="w", delete=False) as file:
         abs_path = file.name
         # Adding a comment to the file so that Slurm does not display warnings in case of empty file
         file.write("# This file is automatically generated by ParallelCluster\n\n")
-        if dummy_nodes_count > 0:
-            file.write(
-                "NodeName=dummy-compute[1-{0}] CPUs={1} State=FUTURE\n".format(
-                    dummy_nodes_count, instance_properties["slots"]
-                )
-            )
-        for node in node_list:
+        for node in generic_node_list:
             file.write("{0}".format(node))
 
     # Update permissions on new file
     os.chmod(abs_path, 0o744)
     # Move new file
-    move(abs_path, PCLUSTER_NODES_CONFIG)
+    move(abs_path, config_file_path)
 
 
-def update_cluster(max_cluster_size, cluster_user, update_events, instance_properties):
+def update_gres_node_lists(update_events):
+    """
+    Process events and add new gres nodes to existing gres_node_list.
+
+    Only returns updated gres_node_list; nodes_to_restart is handled by update_node_lists function.
+    """
+    gres_node_list = _read_node_list(PCLUSTER_GRES_CONFIG)
+    for event in update_events:
+        if event.action == "REMOVE":
+            node_name = "NodeName={0}".format(event.host.hostname)
+            gres_node_list = [node for node in gres_node_list if node.split()[0] != node_name]
+        elif event.action == "ADD":
+            # Only add new node in gres.conf if instance has GPU
+            if int(event.host.gpus) != 0:
+                new_gres_node = "NodeName={nodename} Name=gpu Type=tesla File=/dev/nvidia[0-{gpus}]\n".format(
+                    nodename=event.host.hostname, gpus=int(event.host.gpus) - 1
+                )
+
+                if new_gres_node not in gres_node_list:
+                    gres_node_list.append(new_gres_node)
+
+    return gres_node_list
+
+
+def update_node_lists(update_events):
+    """
+    Process events and add new nodes to existing node_list.
+
+    Returns updated node_list and nodes_to_restart.
+    Since events are the same, nodes_to_restart should be the same for gres_nodes and nodes.
+    """
     # Get the current node list
-    node_list = _read_node_list()
+    node_list = _read_node_list(PCLUSTER_NODES_CONFIG)
     nodes_to_restart = []
     for event in update_events:
         if event.action == "REMOVE":
             node_name = "NodeName={0}".format(event.host.hostname)
             node_list = [node for node in node_list if node.split()[0] != node_name]
         elif event.action == "ADD":
-            # Add new node
-            new_node = "NodeName={nodename} CPUs={cpus} State=UNKNOWN\n".format(
-                nodename=event.host.hostname, cpus=event.host.slots
+            # Only include GPU info if instance has GPU
+            gpu_info = "Gres=gpu:tesla:{gpus} ".format(gpus=event.host.gpus) if int(event.host.gpus) != 0 else ""
```

Comment:
can we have gpus stored already as an int in host?

Reply:
setting gpus as int in `get_instance_properties`